### PR TITLE
Add Python dependency info to packages and display it in DPM UI

### PR DESCRIPTION
**Pull request by zeusongit**
### Purpose
https://jira.autodesk.com/browse/DYN-3643

- Adds a method to scan incoming packages for python nodes and dependent packages
- Adds python engine dependency to package metadata.
- Adds UI component to display dependencies by version
- Updated UI to display latest version's dependencies

PS: One time process to scan and update all existing packages is under development.

SS:
Screen Shot 2021-06-03 at 12 38 40 AM

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@DynamoDS/dynamo

### CODE REVIEW

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -77,6 +77,65 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+            return null;
+        }
+
+        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            return null;
+        }
+    }
+
+    public class EmptyDepStringToCollapsedConverter : IValueConverter
```
Comment (QilongTang) [inline src/DynamoCoreWpf/UI/Converters.cs:112]: Let's add some comments for public class

File: src/DynamoCoreWpf/ViewModels/PackageManager/PublishPackageViewModel.cs [hunk 2]
```diff
@@ -1349,14 +1362,23 @@ private IEnumerable<string> BuildPackage()
...
                 Package.Dependencies.Clear();
                 GetAllDependencies().ToList().ForEach(Package.Dependencies.Add);
 
+                Package.HostDependencies = Enumerable.Empty<string>();
+                Package.HostDependencies = SelectedHosts;
+                foreach (var py in GetPythonDependency()) 
+                {
+                    if (!Package.HostDependencies.Contains(py)) 
+                    {
+                        Package.HostDependencies = Package.HostDependencies.Concat(new[] { py });
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PublishPackageViewModel.cs:1378]: Would you remind me does this `HostDependencies` get saved into DB?

Comment (zeusongit) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PublishPackageViewModel.cs:1378]: yes

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj [hunk 3]
```diff
@@ -1255,6 +1255,10 @@
       <Project>{6e0a079e-85f1-45a1-ad5b-9855e4344809}</Project>
       <Name>Units</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Libraries\PythonNodeModels\PythonNodeModels.csproj">
+      <Project>{8872CA17-C10D-43B9-8393-5C5A57065EB0}</Project>
+      <Name>PythonNodeModels</Name>
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: While I don't *think* this causes a circular dependency, it is a bit odd to have DynamoCoreWPF depend on a node library.

 If we could figure out a way to solve this with reflection or some other mechanism - for example an interface in DynamoServiecs that might better decouple the core from our libraries and make it clear what the required interfaces actually are.

Comment (mjkkirschner) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: On the other hand, one could make the argument that this is now a core requirement for a UI feature and as such we need to have this reference.

You may want to spend more time generalizing this if we still intend to allow for fully dynamic engine loading in the future - if not, then maybe defining those interfaces is not worth it at the moment.

Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: @mjkkirschner I feel the problem can be traced back to why package manager UI is part of DynamoCoreWpf. To me it seems PM should be more a view extension but not the current way that UI in CoreWpf but data model as extension.. On that note, if the original design is that PM should be part of Core then I feel this dependency is accurate.. Or we could move Python loading to DynamoCore but that also seems not necessary.

**Comment by mjkkirschner** [comment]
@QilongTang - there is actually a PackageManagerUIViewExtension - we just never finished moving the UI out of core and into the view extension :) - more tech tebt.

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj (hunk 3 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: @QilongTang - there is actually a PackageManagerUIViewExtension - we just never finished moving the UI out of core and into the view extension :) - more tech debt.

Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: Ah.. @mjkkirschner Then are you fine of this temp dependency and we shall get to it. Our team have Package Manager UI refresh this quarter and it seems the tech debt can be address in that epic?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: sounds good - maybe we just make a jira task for this (python references) and another for the packageManager UI refactor tech debt to track both of them?

Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: Will do @mjkkirschner

**Review by QilongTang — APPROVED** [review]

**Review by QilongTang — APPROVED** [review]

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj (hunk 3 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:1260]: Logged as https://jira.autodesk.com/browse/DYN-3765

**Comment by mjkkirschner** [comment]
@zeusongit @QilongTang - I see the tests are failing - my guess is because some python assemblies are being copied to multiple folders now because of the new references.

mark it private false or add a cleanup target to one of the scripts we have to delete those binaries.

**Pull request merged**